compiler/test/suites/basic_functionality.re

```diff
@@ -377,6 +377,6 @@ describe("basic functionality", ({test, testSkip}) => {
     ~config_fn=smallestFileConfig,
     "smallest_grain_program",
     "",
-    4769,
+    6424,
```

Comment:
Why did this go up?

Is it because of names changing and debug info?

Reply:
I meant to leave a comment about this; thanks for pointing it out. The old linker had a concept of hard/soft dependencies. If a file was a dependency but no values/functions from that module were actually uses (just types or an unused import), then the linker would compile in type information but not link in the actual files.

At first thought this seems like good behavior, but if this is done, side effects from importing those modules don't happen, which is bad. For example, a normal Grain program that doesn't use anything in `Pervasives` doesn't link in `Pervasives`, which means the `setupExceptions` side effects don't happen, and throwing exceptions in those programs prints `GrainError` instead of the actual error. In practice, users don't run into this because most programs use something from Pervasives.

This new linker links in all dependencies, which is more correct. As I mentioned, in practice this doesn't make a real difference for most users. Side note: the major reason the original implementation did this was because the compiler wasn't great at removing unused code from the resulting binary and this resulted in large module sizes. The compiler has gotten much better and this isn't really an issue anymore.

I also discovered a few more configurations producing smaller programs while working on this, as small as 25 bytes.

Reply:
Ah okay then this probably closes #1903 then.